pipeline/backend/lxc/lxc.go

```diff
@@ -0,0 +1,286 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lxc
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/woodpecker-ci/woodpecker/pipeline/backend/types"
+	"github.com/woodpecker-ci/woodpecker/shared/constant"
+)
+
+var scriptTemplate = template.Must(template.New("script").Parse(`#!/bin/sh -xe
+lxc-create --name="{{.Name}}" --template={{.Template}} -- --release {{.Release}} $packages
+tee -a /var/lib/lxc/{{.Name}}/config <<'EOF'                                                                                                                   
+security.nesting = true
+lxc.cap.drop =
+lxc.apparmor.profile = unconfined
+EOF
+
+mkdir /var/lib/lxc/{{.Name}}/rootfs/woodpecker
+mount --bind {{.Workspace}} /var/lib/lxc/{{.Name}}/rootfs/woodpecker
+
+mkdir /var/lib/lxc/{{.Name}}/rootfs/rundir
+mount --bind {{.RunDir}} /var/lib/lxc/{{.Name}}/rootfs/rundir
+
+lxc-start {{.Name}}
+lxc-wait --name {{.Name}} --state RUNNING
+lxc-attach --name {{.Name}} -- /rundir/networking.sh
+lxc-attach --name {{.Name}} -- /bin/sh -c 'cd "/woodpecker/{{ .Repo }}" && /bin/sh -ex /rundir/{{ .Script }}'
+`))
+
+func writeScript(t *template.Template, config any, script string) error {
+	f, err := os.Create(script)
+	if err != nil {
+		return err
+	}
+	if err := os.Chmod(script, 0o755); err != nil {
+		return err
+	}
+	if err := t.Execute(f, config); err != nil {
+		return err
+	}
+	if err := f.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+type lxc struct {
+	cmd       *exec.Cmd
+	output    io.ReadCloser
+	rundir    string
+	workspace string
+	name      string
+}
+
+// make sure lxc implements Engine
+var _ types.Engine = &lxc{}
+
+// New returns a new lxc Engine.
+func New() types.Engine {
+	return &lxc{}
+}
+
+func (e *lxc) Name() string {
+	return "lxc"
+}
+
+func (e *lxc) IsAvailable() bool {
+	return true
+}
+
+func (e *lxc) Load() error {
+	dir, err := os.MkdirTemp("", "woodpecker-lxc-*")
+	e.rundir = dir
+	e.name = path.Base(dir)
+	return err
+}
+
+var serviceHostnameTemplate = template.Must(template.New("hostnames").Parse(`#!/bin/sh -ex
+#
+# Wait until service containers get an IP and set /etc/hosts with their name
+#
+cat /rundir/service-alias | while read name alias ; do
+  for d in $(seq 60); do
+    getent hosts $name > /dev/null && break
+    sleep 1
+  done
+  echo $(getent hosts $name) $alias
+done | tee -a /etc/hosts
+#
+# Wait until internet connectivity is ready
+#
+for d in $(seq 60); do
+  getent hosts wikipedia.org > /dev/null && break
+  sleep 1
+done
+getent hosts wikipedia.org
```

Comment:
Is this a smart choice to use wikipedia, I understand that they have a really small chance to be down but what will append if the system administrator intentionally disable access to the outside world ?
It is possible to want a localnet only container, for some reason.

Reply:
I don't know how to do the same reliably using a different method. I'm open to ideas.

Reply:
Maybe you can set some sort of `AGENT_PING_HOST` variable on the agent, allowing the wikipedia.org host to be overwritten by the agent admin.

Reply:
That would work :+1:

Reply:
Disable it in LXC, and set the ping address to localhost. That should work.

Reply:
WOODPECKER_BACKEND_LXC_NETWORK_READY_HOST was added